# <center> Modifier le script python </center>

## Etapes par étapes
Si vous souhaitez étudier une caractéristique autre que le nombre de branches cela est possible. Avant de commencer le code il est nécessaire d'importer les bibliothèques que nous souhaitons utiliser dans notre cas la bibliothèque pandas avec la ligne de code *import pandas as pd*, ici *as* permet de créér un alias au lieu de noter par la suite *pandas* il suffit de noter *pd*. Importer également les bibliothèques *os* et *glob*.

In [48]:
import pandas as pd
import os
import glob

Ensuite nous souhaitons ouvrir un fichier au format csv il faut utiliser la fonction **read_csv** et lui donner le chemin du fichier au format csv à ouvrir. Ici la variable chemin contient le chemin du fichier au format csv. L'ajout du *r* avant le chemin et la méthode *replace("\\","/")* après le chemin permettent de remplacer les antislashs (\\) en slashs (/).

In [8]:
chemin = r"C:\Users\naima.ammiche\Desktop\Données - traitement images\Classification plaque 1\A01_save.csv".replace("\\","/")
data = pd.read_csv(chemin)
data

# Branches  # Junctions  # End-point voxels  # Junction voxels  \
0      1         130           63                  35                320   
1      2          23           11                   9                 59   
2      3          54           27                  21                113   
3      4          34           16                  16                 68   
4      5         120           61                  44                233   
...   ..         ...          ...                 ...                ...   
2002  44         181           83                  54                433   
2003  45          67           35                  20                135   
2004  46          37           15                  13                109   
2005  47          45           13                  18                177   
2006  48          71           36                  27                146   

      # Slab voxels  Average Branch Length  # Triple points  \
0               827                   4.96               40   
1               261                   7.77                8   
2               508                   6.37               21   
3               594                  10.61               12   
4               907                   5.40               51   
...             ...                    ...              ...   
2002            913                   4.52               50   
2003            926                   8.77               29   
2004            122                   4.05                5   
2005             98                   4.35                4   
2006            358                   4.00               31   

      # Quadruple points  Maximum Branch Length  Longest Shortest Path  \
0                     15                  24.28                 168.01   
1                      2                  29.55                 131.61   
2                      6                  46.46                 136.50   
3                      4                  68.75                 186.90   
4                      6                  16.44                 102.87   
...                  ...                    ...                    ...   
2002                  21                  18.82                 142.74   
2003                   5                  25.07                 113.68   
2004                   5                   7.83                  38.79   
2005                   5                  15.79                  45.92   
2006                   3                  12.10                  98.62   

         spx     spy  spz  
0      64.47  148.00    0  
1      65.83  500.31    0  
2     101.24  130.30    0  
3     269.22  131.66    0  
4     243.80  492.59    0  
...      ...     ...  ...  
2002  482.15  376.82    0  
2003  499.85  476.25    0  
2004  458.09  186.14    0  
2005  483.06  123.94    0  
2006  492.14  148.91    0  

[2007 rows x 14 columns]

Le data frame ci-dessus est enregistré dans la variable *data* et pour l'afficher il suffit de mettre *data* ou bien *print(data)*. Une ligne correspond à une cellule d'oligodendrocytes présente dans le puits et en colonne les caractéristiques associés. Pour avoir les statistiques descriptives de chaque variable il suffit d'appeler la méthode *describe()* sur se data frame. Vous aurez le **nombre total de ligne, moyenne, écart-type, minimum, quartiles et maximum** pour chaque variable du data frame.

In [10]:
data.describe()

# Branches  # Junctions  # End-point voxels  \
count  2007.000000  2007.000000  2007.000000         2007.000000   
mean     21.989038   100.903837    49.785750           33.378176   
std      14.733724    65.590114    33.795378           18.696212   
min       1.000000    12.000000     6.000000            5.000000   
25%      10.000000    53.000000    25.000000           20.000000   
50%      20.000000    86.000000    42.000000           30.000000   
75%      32.000000   130.500000    65.000000           42.000000   
max      76.000000   572.000000   296.000000          146.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        2007.000000    2007.000000            2007.000000   
mean          218.686597     673.425511               5.278819   
std           156.571454     418.466577               1.072032   
min            22.000000      79.000000               3.010000   
25%           111.000000     358.000000               4.530000   
50%           177.000000     598.000000               5.090000   
75%           283.000000     888.000000               5.830000   
max          1335.000000    3992.000000              14.140000   

       # Triple points  # Quadruple points  Maximum Branch Length  \
count      2007.000000         2007.000000             2007.00000   
mean         37.555556            8.269557               19.53998   
std          25.754260            6.671511                6.87098   
min           2.000000            0.000000                6.28000   
25%          19.000000            4.000000               15.08500   
50%          32.000000            6.000000               18.27000   
75%          49.000000           11.000000               22.60000   
max         217.000000           56.000000               68.75000   

       Longest Shortest Path          spx          spy     spz  
count            2007.000000  2007.000000  2007.000000  2007.0  
mean              105.498859   288.676532   250.064649     0.0  
std                36.461241   144.668688   145.492005     0.0  
min                33.170000     0.000000     0.000000     0.0  
25%                81.395000   164.120000   123.940000     0.0  
50%               101.330000   302.360000   251.970000     0.0  
75%               124.040000   416.090000   371.830000     0.0  
max               297.980000   500.760000   500.760000     0.0

Si vous souhaitez récupérer uniquement la variable **# Junctions**, il faut apeller le data frame et mettre entre crochets le nom de la variable.

In [12]:
data["# Junctions"]

0       63
1       11
2       27
3       16
4       61
        ..
2002    83
2003    35
2004    15
2005    13
2006    36
Name: # Junctions, Length: 2007, dtype: int64

La cellule ci-dessus représente en lignes les cellules d'oligodendrocytes et en colonne le nombre de jonction. *Exemple la cellule 0 possède 63 jonctions, la cellule 1 possède 11 jonctions etc.* Pour récupérer uniquement les jonctions inférieures à 33  ajouter une condition.

Les différents types de conditions possibles pour le data frame :
- **>** pour supérieur, **>=** supérieur ou égal
- **<** inférieur, **<=** inférieur ou égal
- **&** signifie **ET** pour combinées deux conditions *exemple supérieur à 30 et inférieur à 35 :  **>30 & <35** *
- **|** signifie **Ou** sélectionne lorsque l'une des conditions est valide *exemple inférieur à 10 ou supérieur à 60 : **<10 | > 60** * 

In [26]:
#Jonction inférieur à 33
data[data["# Junctions"] < 33]

# Branches  # Junctions  # End-point voxels  # Junction voxels  \
1      2          23           11                   9                 59   
2      3          54           27                  21                113   
3      4          34           16                  16                 68   
5      6          73           31                  26                156   
6      7          41           22                  14                 70   
...   ..         ...          ...                 ...                ...   
1998  40          43           20                  19                 77   
1999  41          57           22                  18                159   
2001  43          54           28                  19                105   
2004  46          37           15                  13                109   
2005  47          45           13                  18                177   

      # Slab voxels  Average Branch Length  # Triple points  \
1               261                   7.77                8   
2               508                   6.37               21   
3               594                  10.61               12   
5               194                   3.09               18   
6               335                   5.50               20   
...             ...                    ...              ...   
1998            451                   7.02               15   
1999            243                   4.70               11   
2001            525                   6.64               22   
2004            122                   4.05                5   
2005             98                   4.35                4   

      # Quadruple points  Maximum Branch Length  Longest Shortest Path  \
1                      2                  29.55                 131.61   
2                      6                  46.46                 136.50   
3                      4                  68.75                 186.90   
5                      6                   7.86                  66.57   
6                      2                  14.55                  68.01   
...                  ...                    ...                    ...   
1998                   3                  25.43                  93.57   
1999                   5                  13.15                  62.76   
2001                   4                  17.27                 117.78   
2004                   5                   7.83                  38.79   
2005                   5                  15.79                  45.92   

         spx     spy  spz  
1      65.83  500.31    0  
2     101.24  130.30    0  
3     269.22  131.66    0  
5     280.12  424.04    0  
6     298.28  471.71    0  
...      ...     ...  ...  
1998  422.22  281.93    0  
1999  403.15  458.54    0  
2001  459.90  285.11    0  
2004  458.09  186.14    0  
2005  483.06  123.94    0  

[733 rows x 14 columns]

Nous venons de récupérer uniquement les cellules avec les jonctions inférieures à 33, pour savoir le nombre de cellules il suffit d'avoir le nombre de lignes et donc utiliser la méthode *shape* représentant les dimensions. (lignes, colonne) pour avoir uniquement les lignes nous récupérons uniquement le premier élement du tuple donc *shape[0]*

In [25]:
oligo_inf33 = data[data["# Junctions"] < 33].shape[0]
oligo_inf33

733

Nous avons dans le puits A01 de la plaque1, 733 oligodendrocytes avec moins de 33 jonctions pour obtenir le pourcentage il faut savoir combien d'oligodendrocytes sont présents dans ce puits avec la même méthode *shape[0]* mais appliquer sur l'ensemble du data frame avant la sélection du nombre de jonctions.

In [18]:
oligo_total = data.shape[0]
oligo_total

2007

In [20]:
#Calcul du pourcentage
(oligo_inf33/oligo_total)*100

36.52217239661186

Les oligodendrocytes avec moins de 33 jonctions repésentent 36,5 % des oligodendrocytes du puits A01 de la plaque1. 


Si nous avons une liste contenant des noms de fichiers au format csv pour la parcourir il faut utiliser une boucle *for*. Un exemple est présent dans la cellule suivante. La ligne de code **for csv in liste** permet de parcourir les fichiers un par un donc "img1_A01.csv" puis "img2_A01.csv", "img1_A02.csv" et enfin "img2_A02.csv". Pour chaque fichier affiché un message avec *print* le message affiché est **Nouveau fichier** suivi du nom du fichier csv et du puits associé. Ici csv correspond au nom du fichier exemple **img2_A01.csv** qui est une chaîne de caractères, il est possible de récupérer une sous-chaine de caractère exemple :
- csv correspond à **img2_A01.csv**
- csv[0:4] correspond à la sous chaine de caractère 0 à 4 exclus donc récupérer les caractères **img**. Le caractère 0 correspond à **i** le caractère 2 à **m** et le caractère 3 à **g**.
- csv[-1] correspond au premier caractère en partant de la fin (car il y a le symbole **-**) donc il s'agit de **v**
- csv[-7:-4] correspond à **A01** car il part du 7e caractère à partir de la fin donc **A** jusqu'au 4e caractère en partant de la fin exclue donc "**.**" n'est pas pris en compte. 
Cette méthode est ensuite utilisée dans le script pour récupérer le nom du puits qui est en train d'être analysé et concaténer les fichiers d'un même puits.


In [42]:
liste = ["img1_A01.csv","img2_A01.csv","img1_A02.csv","img2_A02.csv"]

for csv in liste:
    print(f"\nNouveau fichier\n{csv}")
    
    # Nom du puits
    puits = csv[-7:-4]
    print(puits)


Nouveau fichier
img1_A01.csv
A01

Nouveau fichier
img2_A01.csv
A01

Nouveau fichier
img1_A02.csv
A02

Nouveau fichier
img2_A02.csv
A02



## Modification du script classification
Nous venons de voir comment récupérer les informations sur un fichier au format csv. Avec la macro de classification des oligodendrocytes nous avons plusieurs fichiers au format csv. Il faut donner le chemin du répertoire contenant l'ensemble des fichiers csv et changer les antislashs en slashs comme vu précédemment.

In [62]:
csv_dir = r"C:\Users\naima.ammiche\Desktop\Données - traitement images\Classification plaqueA".replace("\\","/")

Ensuite récupérer l'ensemble des fichiers csv de ce répertoire et les mettre dans la variable **csv_files** sous forme de liste. Nous parcourons ici la liste **csv_files** les éléments qu'elle contient avec la boucle *for* et la méthode *print*. 

In [63]:
# Récupérer l'ensemble des fichiers csv produit
csv_regex = os.path.join(csv_dir.replace("\\", "/"), "*.csv") # Une chaîne de caractère des nom des fichiers à sélectionner
csv_files = glob.glob(csv_regex) # Liste de l'ensemble des fichiers CSV

# Afficher les élements de la liste
for csv in csv_files:
    print(csv)

C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_A01.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_A04.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_A07.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_A10.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_B01.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_B04.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_B07.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_B10.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_C01.csv
C:/Users/naima.ammiche/Desktop/Données - traitement images/Classification plaqueA\img0_C04.csv
C:/Users/naima.ammiche/Desktop/Données - traitemen

Nous pouvons voir que cette liste contient les chemins de chaque fichier csv qui a été obtenu avec la macro de classification d'oligodendrocytes. Nous venons donc de récupérer le chemin de tous les fichiers au format csv dans la liste **csv_files**. Parcourir la liste **csv_files** avec une boucle *for* et pour chaque fichier récupérer le nom du puits comme expliqué précédemment. Pour chaque nom de puits récupérer l'ajouter à la liste all_puits avec la méthode *append* et afficher un message avec l'ensemble de puits.

In [64]:
# Liste vide
all_puits = []

# Parcourt le répertoire et pour chaque fichier 
for csv in csv_files:
    # Parcourir csv et ajouter à la liste chaque puits ex [A01, A02, B01 ...]
    puits = csv[-7:-4] 
    if puits not in all_puits:
        all_puits.append(puits) 

# Afficher la liste contenant l'ensemble des puits à analyser
print(f"L'ensemble des puits {all_puits}")

L'ensemble des puits ['A01', 'A04', 'A07', 'A10', 'B01', 'B04', 'B07', 'B10', 'C01', 'C04', 'C07', 'C10', 'D01', 'D04', 'D07', 'D10', 'E01', 'E04', 'E07', 'E10', 'F01', 'F04', 'F07', 'F10', 'G01', 'G04', 'G07', 'G10', 'H01', 'H04', 'H07', 'H10']


Parcourir ensuite la liste **all_puits** représentant l'ensemble des puits donc A01 puis A02 ... et enfin H10. Pour chaque puits parcourut par exemple lorsqu'on parcourt A01  on parcourt la liste **csv_files** (la liste de tous les fichiers csv). 

Vérifier si le nom du puits ici parcourut de la liste **all_puits** exemple A01 et celui présent sur le nom du fichier csv parcourut de la liste **csv_files**  exemple img1_A01.csv sont idendiques avec la ligne de code **if name_puits == name:** dans ce cas essayer d'ouvrir ce fichier avec **try** s'il est vide un message apparaît notamment le message présent dans **except**.

Ensuite la ligne de code **if csv == csv_files[0]:** permet de vérifier si le fichier parcourut actuellement exemple **img1_A01** correspond au premier élément de la liste **csv_files** dans ce cas il produit un data frame **df** dans le cas contraire il l'ajoutait au data frame existant avec la ligne **df = pd.concat(objs = [cluster_skeleton,df], axis =0)**. Le fichier csv exemple *img1_A01.csv* est supprimée après avoir été ajouté au data frame. A la fin de la boucle donc lorsque tous les fichiers csv sont parcourus de la liste **csv_files**, le data frame **df** est enregistré au format csv dont le nom est par exemple **A01_save** pour le puits A01.

La ligne de code **print(f"Puits {name}\n{df.describe()}\n\n")** permet d'afficher le nom du puits suivi du data frame associé après avoir concaténé. La fonction describe() permet comme expliqué plus haut d'obtenir les statistiques du data frame. Les symboles **\\n** signifient le retour à la ligne lors de l'affichage pour avoir une meilleure visibilité du message.

Enfin la dernière ligne de code **df.drop(df.index[::], inplace=True)** permet de supprimer le data frame df après avoir était enregistrer au format csv pour produire un nouveau data frame pour un autre puits de la liste **all_puits**. *Exemple après avoir enregistré A01_save.csv le data df est supprimé et le même processus est reproduit pour le puits A04 et concaténé les fichiers csv de ce même puits et ce jusqu'au dernier puits H10*.

In [65]:
# Parcourt la liste des puits A01, A02, B01... 
for name in all_puits:
    
    # Parcourt les fichiers csv obtenus par la macro de classification des oligodendrocytes
    for csv in csv_files:
        name_puits = csv[-7:-4]
        
        if name_puits == name:                
            # Essayer d'ouvrir le fichier CSV en data frame et affiché un message lorsque le fichier est vide 
            # (probablement causé lorsque des reflets présent sur l'image empêche l'analyse)
            try:  
                cluster_skeleton = pd.read_csv(csv) 
            except:
                print(f"Ce fichier est vide :  {csv}") 
    
            # Si c'est le prermier fichier CSV crée un data frame, sinon l'ajouter au dataframe existant
            if csv == csv_files[0]:
                df = cluster_skeleton
            else :
                df = pd.concat(objs = [cluster_skeleton,df], axis =0)
            os.remove(csv) 

    # Enregistrer le fichier data contenant les résultats de l'ensemble du puits au format CSV
    df.to_csv(os.path.join(csv_dir.replace("\\", "/"), name+"_save.csv"), index = False) # Sauvegarder les résultats dans un seul fichier csv
    df = pd.read_csv(csv_dir.replace("\\", "/")+"\\"+name+"_save.csv") # Ouvrir le fichier csv à analyser # Ouvrir un fichier CSV seulement
    print(f"Puits {name}\n{df.describe()}\n\n")
    
    # Supprimer le dataframe, vous pouvez retrouver le data frame le fichier csv sauvegarder
    df.drop(df.index[::], inplace=True) 

Puits A01
                     # Branches  # Junctions  # End-point voxels  \
count  3781.000000  3781.000000  3781.000000         3781.000000   
mean     43.736049   101.139646    33.088601           35.599577   
std      37.132298    88.892867    23.963902           32.367123   
min       1.000000    18.000000     1.000000            1.000000   
25%      17.000000    50.000000    17.000000           18.000000   
50%      36.000000    73.000000    26.000000           26.000000   
75%      59.000000   117.000000    41.000000           40.000000   
max     215.000000   936.000000   219.000000          417.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        3781.000000    3781.000000            3781.000000   
mean          382.516530     303.434541               5.703303   
std           496.003857     213.783860               4.136235   
min            26.000000      34.000000               2.750000   
25%           141.000000     157.000000        

Puits B01
                     # Branches  # Junctions  # End-point voxels  \
count  4170.000000  4170.000000  4170.000000         4170.000000   
mean     38.984892    74.358034    31.913909           25.133573   
std      25.312497    41.208182    18.013082           13.094492   
min       1.000000    18.000000     2.000000            4.000000   
25%      18.000000    46.000000    19.000000           16.000000   
50%      36.000000    62.000000    27.000000           22.000000   
75%      57.000000    91.000000    40.000000           31.000000   
max     119.000000   429.000000   171.000000          124.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        4170.000000    4170.000000            4170.000000   
mean          200.520624     306.486571               4.359228   
std           130.069744     167.814214               0.761270   
min            22.000000      48.000000               3.050000   
25%           114.000000     189.000000        

Puits C01
                     # Branches  # Junctions  # End-point voxels  \
count  4624.000000  4624.000000  4624.000000         4624.000000   
mean     49.636678   143.414360    40.410251           51.801254   
std      39.110878   156.267238    30.207727           58.560389   
min       1.000000    20.000000     3.000000            5.000000   
25%      20.000000    57.000000    20.000000           21.000000   
50%      40.000000    87.000000    31.000000           31.000000   
75%      68.250000   157.000000    52.000000           54.000000   
max     177.000000  1625.000000   323.000000          647.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        4624.000000    4624.000000            4624.000000   
mean          572.332180     350.057310               6.060634   
std           822.232739     253.588688               3.819917   
min            39.000000      32.000000               2.680000   
25%           160.000000     174.000000        

Puits D01
                     # Branches  # Junctions  # End-point voxels  \
count  4151.000000  4151.000000  4151.000000         4151.000000   
mean     37.532643    83.120453    33.585160           28.231510   
std      23.286096    50.076741    20.917299           15.996269   
min       1.000000    13.000000     4.000000            1.000000   
25%      18.000000    50.000000    20.000000           18.000000   
50%      36.000000    69.000000    28.000000           24.000000   
75%      55.000000   101.000000    42.000000           35.000000   
max     103.000000  1146.000000   482.000000          331.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        4151.000000    4151.000000            4151.000000   
mean          240.840761     287.411949               4.232103   
std           159.130897     173.905922               0.871668   
min            35.000000      53.000000               2.820000   
25%           139.000000     172.500000        

Puits E01
                     # Branches  # Junctions  # End-point voxels  \
count  4507.000000  4507.000000  4507.000000         4507.000000   
mean     39.741513    83.466607    34.747060           28.235190   
std      24.394959    45.707463    19.669899           14.890343   
min       1.000000    21.000000     4.000000            5.000000   
25%      19.000000    51.000000    21.000000           18.000000   
50%      38.000000    71.000000    29.000000           25.000000   
75%      58.000000   103.000000    44.000000           35.000000   
max     106.000000   421.000000   173.000000          157.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        4507.000000    4507.000000            4507.000000   
mean          234.520967     309.780342               4.249410   
std           142.768865     172.720296               0.682174   
min            41.000000      48.000000               2.810000   
25%           136.000000     186.000000        

Puits F01
                     # Branches  # Junctions  # End-point voxels  \
count  3249.000000  3249.000000  3249.000000         3249.000000   
mean     33.105571    93.716528    37.777778           31.799938   
std      23.000514    57.346470    24.918955           18.372161   
min       1.000000    25.000000     2.000000            5.000000   
25%      14.000000    53.000000    20.000000           19.000000   
50%      29.000000    77.000000    31.000000           27.000000   
75%      49.000000   116.000000    48.000000           39.000000   
max     106.000000   571.000000   275.000000          185.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        3249.000000     3249.00000            3249.000000   
mean          274.130194      331.44506               4.356691   
std           178.951728      229.19703               0.901330   
min            37.000000       40.00000               2.640000   
25%           152.000000      166.00000        

Puits G01
                     # Branches  # Junctions  # End-point voxels  \
count  5084.000000  5084.000000  5084.000000         5084.000000   
mean     43.723446    93.821794    42.727773           31.012392   
std      25.970568    56.945250    27.178012           16.074905   
min       1.000000    19.000000     5.000000            5.000000   
25%      22.000000    53.000000    23.000000           19.000000   
50%      43.000000    79.000000    36.000000           27.000000   
75%      64.000000   119.000000    55.000000           39.000000   
max     123.000000   615.000000   303.000000          181.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        5084.000000    5084.000000            5084.000000   
mean          235.004917     438.277537               4.440565   
std           161.483229     259.066605               0.618433   
min            29.000000      47.000000               2.850000   
25%           125.000000     248.000000        

Puits H01
                     # Branches  # Junctions  # End-point voxels  \
count  6300.000000  6300.000000  6300.000000         6300.000000   
mean     55.232063   133.236984    39.683333           44.135714   
std      33.424309   149.648596    31.863600           52.432657   
min       1.000000    19.000000     1.000000            4.000000   
25%      27.000000    59.000000    20.000000           21.000000   
50%      54.000000    89.000000    31.000000           30.000000   
75%      81.000000   149.000000    50.000000           48.000000   
max     156.000000  2185.000000   426.000000          834.000000   

       # Junction voxels  # Slab voxels  Average Branch Length  \
count        6300.000000    6300.000000            6300.000000   
mean          565.878730     321.731111               5.977113   
std           909.121204     248.840459               6.215946   
min            54.000000      30.000000               2.830000   
25%           190.000000     162.000000        

Avant de commencer l'analyse, initialiser les variables représentant ici les liste par exemple si on veut 3 classes pour les jonctions avec moins de 20 jonctions, entre 20 et 30 et supérieur à 30 alors il faut 3 listes vide vous pouvez leur donner le nom de variable que vous souhaitez.  La variable index_classification contiendra les puits analysés.  

In [66]:
# Initialisation
inf20 = []
entre20_30 = []
sup30 = []

# Légende: les puits analysé
index_classification = []

Les fichiers sont concaténés nous avons ainsi un fichier csv par puits qui regroupe en lignes les oligodendrocytes du puits et en colonne les différentes caractéristiques. Le même processus que précédement est réalisé c'est-à-dire récupérer les fichiers au format csv mais cette for il ne s'agit pas des fichiers dont le nom est *img1_A01.csv, img2_A01.csv, ... img1_H10.csv, img100_H10 etc.* mais des fichiers obtenus avec la concaténation donc *A01_save.csv, A04_save.csv ... H10_save.csv*

Parcourir la liste **csv_files** qui contient les fichiers *A01_save.csv, A04_save.csv ... H10_save.csv* et récupérer le nom du puits pour chaque fichier et l'ajouter à la liste **index_classification**. Ensuite pour chaque fichier csv essayer de l'ouvrir en data frame avec la ligne de code **pd.read_csv(csv)** dans le cas contraire afficher un message pour prévenir que le fichier est vide

Ensuite à partir du commentaire *# Caractéristique* vous pouvez modifier mettre autant de clusters que nécessaire par exemple 3 clusters dans cet exemple et pour chaque cluster récupéré les variables souhaiter inférieur à 2 jonctione <20, entre 20 et 30 jonctions >=20 & <30 et supérieur à 30 jonctions >=30.

Calculer ensuite le pourcentage d'oligodendrocytes pour chaque cluster. Ainsi les variables **value** contiennent un pourcentage, *exemple ici value1 contient le pourcentage du cluster 1 et donc du nombre de jonctions inférieur à 20*. L'étape suivante consiste à ajouter ces pourcentages aux listes correspondantes donc ajouter *value* à la liste *inf20* avec la ligne de code **inf20.append(value1)**. Cette ligne de code prend la liste **inf20** et ajoute la valeur **value1** à la fin de la liste. 

Enfin un affichage permet de voir pour chaque puits le pourcentage pour chaque cluster. La partie affichage comporte les lignes de code : 
- **print(f"\nPuits {puits} : \nDimension totale du data frame : {df.shape} : Il y a {df.shape[0]} oligodendrocytes dans ce puits. Il y a {df.shape[1]} caractéristiques étudiées sur ces oligodendrocytes")** qui affiche le nom du puits et les dimensions du data frame en précisant le nombre d'oligodendrocytes (=nombre de ligne) et le nombre de caractéristiques (=nombre de colonne).
- **print(f"<20 : {value1:.2f} %\nentre 20 et 30 : {value2:.2f}%\>=30 : {value3:.2f}%")** qui affiche le pourcentage de chaque cluster

In [67]:
# Récupérer l'ensemble des fichiers csv dans la liste csv_files
csv_regex = os.path.join(csv_dir.replace("\\", "/"), "*.csv") # Une chaîne de caractère contenant nom des fichiers à sélectionner
csv_files = glob.glob(csv_regex) # Liste de l'ensemble des fichiers CSV

# Parcourt les fichiers csv produit à l'étape précèdente
for csv in csv_files:
    
    # Nom du puits
    puits = csv[-12:-9] 
    # Ajouter le puits analyser
    index_classification.append(puits)
    
    # Ouvrir le fichier csv si celui ci est vide un message apparaît
    try :
        df = pd.read_csv(csv)
    except:
        print(f"Ce fichier est vide :  {csv}") 
                
    # Caractéristique
    cluster1 = df[df['# Junctions'] < 20]
    cluster2 = df[(df['# Junctions'] >= 20) & (df['# Junctions'] < 30)]
    cluster3 = df[df['# Junctions'] > 30]
    
    
    # % d'oligodendrocytes pour chaque classe
    oligo_total = df.shape[0] # Nombre total d'oligodendrocytes dans le puits
    value1 = (cluster1.shape[0]/oligo_total)*100 # avec moins de 20  jonctions
    value2 = (cluster2.shape[0]/oligo_total)*100 # entre 150 et 300 branches
    value3 = (cluster3.shape[0]/oligo_total)*100 # entre 300 et 450 branches

    
    # Ajouter ces pourcentage aux listes correspondantes
    inf20.append(value1)
    entre20_30.append(value2)
    sup30.append(value3)

    # AFFICHAGE
    #Dimension du data frame et pourcentage de chaque classe d'oligodendrocytes
    print(f"\nPuits {puits} : \nDimension totale du data frame : {df.shape} : Il y a {df.shape[0]} oligodendrocytes dans ce puits. Il y a {df.shape[1]} caractéristiques étudiées sur ces oligodendrocytes")
    print(f"Pourcentage d'oligodendrocytes par cluster pour le puits {name}") 
    print(f"<20 : {value1:.2f} %\nentre 20 et 30 : {value2:.2f}%\>=30 : {value3:.2f}%")
    
    # Affichage
    df.drop(df.index[::], inplace=True) # Supprimer le dataframe


Puits A01 : 
Dimension totale du data frame : (3781, 14) : Il y a 3781 oligodendrocytes dans ce puits. Il y a 14 caractéristiques étudiées sur ces oligodendrocytes
Pourcentage d'oligodendrocytes par cluster pour le puits H10
<20 : 32.66 %
entre 20 et 30 : 24.31%\>=30 : 40.94%

Puits A04 : 
Dimension totale du data frame : (3631, 14) : Il y a 3631 oligodendrocytes dans ce puits. Il y a 14 caractéristiques étudiées sur ces oligodendrocytes
Pourcentage d'oligodendrocytes par cluster pour le puits H10
<20 : 26.63 %
entre 20 et 30 : 29.61%\>=30 : 41.72%

Puits A07 : 
Dimension totale du data frame : (3960, 14) : Il y a 3960 oligodendrocytes dans ce puits. Il y a 14 caractéristiques étudiées sur ces oligodendrocytes
Pourcentage d'oligodendrocytes par cluster pour le puits H10
<20 : 23.86 %
entre 20 et 30 : 31.01%\>=30 : 42.58%

Puits A10 : 
Dimension totale du data frame : (4195, 14) : Il y a 4195 oligodendrocytes dans ce puits. Il y a 14 caractéristiques étudiées sur ces oligodendrocytes
P

Production du data frame. Choisir le chemin qui contiendra le data frame final et le nom que vous souhaitez donner aux fichiers csv avec les variables **chemin_result et name_result**. 

In [68]:
# Chemin et nom
chemin_result = r"C:\Users\naima.ammiche\Desktop\Données - traitement images".replace("\\","/")
name_result = "Classification_puitsA"

Nous venons de remplir les listes **inf20, entre20_30, sup30 et index_classification**, *exemple ici **inf20** contient le pourcentage d'oligodendrocytes avec moins de 20 jonctions pour chaque puits*. A partir de ces listes utiliser la fonction **pd.DataFrame** et mettre en index la liste **index_classification** qui contiendra en ligne les différents puits. En colonne mettre les listes **inf20, entre20_30 et sup30**. 

In [69]:
# Production du dataframe
data_result_classification = pd.DataFrame({'% <20': inf20,
                   '% entre 20 et 30': entre20_30,
                   '% > 30':sup30,
                  }, index=index_classification)

Utiliser la fonction **to_csv** pour sauvegarder ce data frame au format csv dans le chemin de la variable **chemin_result** et dont le nom du fichier et celui présent dans la variable **name_result**.

In [70]:
# Sauvegarde du data frame au format csv
data_result_classification.to_csv(os.path.join(chemin_result, name_result+".csv")) # A modifier nom et rep

Cette ligne de code permet d'afficher le data frame que vous retrouvez dans le fichier csv.

In [71]:
# Visualisation du data frame
data_result_classification

% <20  % entre 20 et 30     % > 30
A01  32.663317         24.305739  40.941550
A04  26.631782         29.606169  41.724043
A07  23.863636         31.010101  42.575758
A10  32.848629         26.984505  38.665077
B01  25.563549         31.606715  40.143885
B04  33.340588         30.163221  34.341676
B07  29.668737         30.621118  37.287785
B10  32.660210         26.375438  39.109829
C01  23.724048         23.118512  51.362457
C04  24.373402         30.690537  43.069054
C07  24.844720         25.765689  47.590490
C10  21.848739         28.597689  46.691176
D01  23.825584         29.751867  44.543483
D04  30.516628         27.421042  39.866137
D07  24.455611         31.397729  41.950493
D10  29.784138         31.894252  36.308513
E01  21.100510         28.977147  47.459507
E04  28.371977         30.740464  39.042633
E07  25.932367         28.309179  43.033816
E10  27.532777         31.775924  38.331347
F01  22.345337         25.238535  50.538627
F04  25.201072         27.760175  45.137704
F07  20.535069         23.957580  53.555074
F10  23.135056         27.401268  47.367138
G01  15.794650         23.072384  59.225020
G04  20.751851         24.207329  53.028289
G07  23.019026         28.810370  46.038051
G10  19.319862         22.252341  56.727452
H01  23.730159         23.507937  50.809524
H04  26.876640         31.049869  39.317585
H07  26.265365         29.645698  41.757050
H10  30.385200         29.574252  37.506336